rust/arrow/src/compute/kernels/cast.rs

```diff
@@ -313,13 +318,260 @@ pub fn cast(array: &ArrayRef, to_type: &DataType) -> Result<ArrayRef> {
         (Float64, Int64) => cast_numeric_arrays::<Float64Type, Int64Type>(array),
         (Float64, Float32) => cast_numeric_arrays::<Float64Type, Float32Type>(array),
         // end numeric casts
+
+        // temporal casts
+        (Int32, Date32(_)) => cast_array_data::<Date32Type>(array, to_type.clone()),
+        (Date32(_), Int32) => cast_array_data::<Int32Type>(array, to_type.clone()),
+        (Int64, Date64(_)) => cast_array_data::<Date64Type>(array, to_type.clone()),
+        (Date64(_), Int64) => cast_array_data::<Int64Type>(array, to_type.clone()),
+        (Date32(DateUnit::Day), Date64(DateUnit::Millisecond)) => {
+            let date_array = array.as_any().downcast_ref::<Date32Array>().unwrap();
+            let mut b = Date64Builder::new(array.len());
+            for i in 0..array.len() {
+                if array.is_null(i) {
+                    b.append_null()?;
+                } else {
+                    b.append_value(date_array.value(i) as i64 * MILLISECONDS_IN_DAY)?;
+                }
+            }
+
+            Ok(Arc::new(b.finish()) as ArrayRef)
+        }
+        (Date64(DateUnit::Millisecond), Date32(DateUnit::Day)) => {
+            let date_array = array.as_any().downcast_ref::<Date64Array>().unwrap();
+            let mut b = Date32Builder::new(array.len());
+            for i in 0..array.len() {
+                if array.is_null(i) {
+                    b.append_null()?;
+                } else {
+                    b.append_value((date_array.value(i) / MILLISECONDS_IN_DAY) as i32)?;
+                }
+            }
+
+            Ok(Arc::new(b.finish()) as ArrayRef)
+        }
+        (Time32(TimeUnit::Second), Time32(TimeUnit::Millisecond)) => {
+            let time_array = Time32MillisecondArray::from(array.data());
+            let mult =
+                Time32MillisecondArray::from(vec![MILLISECONDS as i32; array.len()]);
+            let time32_ms = multiply(&time_array, &mult)?;
+
+            Ok(Arc::new(time32_ms) as ArrayRef)
+        }
+        (Time32(TimeUnit::Millisecond), Time32(TimeUnit::Second)) => {
+            let time_array = Time32SecondArray::from(array.data());
+            let divisor = Time32SecondArray::from(vec![MILLISECONDS as i32; array.len()]);
+            let time32_s = divide(&time_array, &divisor)?;
+
+            Ok(Arc::new(time32_s) as ArrayRef)
+        }
+        (Time32(from_unit), Time64(to_unit)) => {
+            let time_array = Int32Array::from(array.data());
+            // cast up to i64 so we can multiply
+            let c: Int64Array = numeric_cast(&time_array)?;
```

Comment:
I missed this `numeric_cast`, it might be better for me to loop here like the other places. I'll have a look